feat: api file manager

0santisimo0 opened the pull request:
<!-- Provide a general summary of the pull request in the Title above -->
<!-- Please complete this template before creating the pull request. -->

## Related Tickets & US
- [CS-63 [SP-4]: Implement the API for File Manager](https://app.clickup.com/t/868b038p8)

## Description

<!-- Please include a summary of your changes. -->
<!-- Also include relevant motivation and context. List any dependencies that are required for this. -->

This pull request introduces implementation and tests for the file manager API. The following API methods for the S3 file manager are covered:

- **uploadDirectory(String projectId, String localDir)**: Uploads files from a local directory to an S3 bucket.
- **listFiles(String projectId)**: Lists all the files in an S3 bucket for a specific project.
- **getFileContent(String projectId, String filePath)**: Retrieves the content of a file from an S3 bucket.

Additionally, this PR adds unit tests for these APIs, including:
- Validating successful uploads to S3.
- Handling edge cases like empty directories, invalid paths, and invalid project IDs.
- Simulating exceptions such as IO errors and AWS S3 errors.

improving error handling, and ensuring full test coverage.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

<!-- How was this change tested? -->
<!-- DON'T DELETE THIS SECTION! If no tests added, explain why. -->

This change has been thoroughly tested using JUnit. The following test scenarios were implemented:


- **Api Endpoints tests**: All api endpoints tests and validations
- **Exceptions And Validation tests**: All api Exceptions tests and validations

All tests were executed locally with successful results. 

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

![Captura desde 2024-11-27 11-58-58](https://github.com/user-attachments/assets/41186de4-ae0d-472c-ba0c-4455384110df)

## File Manager API Documentation

This document provides the details of the File Manager API endpoints, including their request formats, responses, and examples.

## Endpoints

### 1. List Files

#### Request
- **Method**: `GET`
- **URL**: `/fileManager/files`

#### Request Parameters
- **projectId**: The unique identifier of the project.

#### Response
- **Status Code**: 200 OK
- **Body**: A JSON array of file names (strings).

#### Example Response
```json
[
    "projects/{projectId}/projectFiles/dirctoryFirst/class1.java",
    "projects/{projectId}/projectFiles/sec/classDirSec.java",
    "projects/{projectId}/projectFiles/subdirector/subClass1.java"
]
```

### 2. Get File Content

#### Request
- **Method**: `GET`
- **URL**: `/fileManager/file`

#### Request Parameters
- **projectId**: The unique identifier of the project.
- **filePath**: The path of the file.

#### Response
- **Status Code**: 200 OK
- **Body**: The content of the file as a string.

#### Example Request
```http
GET /fileManager/file?projectId=test-project-id&filePath=file1.java
```

#### Example Response
```Java
public class Main {
    public static void main(String[] args) {
        System.out.println("Jefferson Viejo");
    }
}
```

### 3. Upload Directory

#### Request
- **Method**: `POST`
- **URL**: `/fileManager/upload`

#### Request Parameters
- **localDir**: The local directory path that contains the files to be uploaded.

#### Response
- **Status Code**: 200 OK
- **Body**: A JSON object that contains the status of each uploaded file.

#### Example Request
```http
POST /fileManager/upload?localDir=test/
```

#### Example Response
```json
{
  "uploadResults": {
    "projects/e011bad2-0b57-4ed3-a278-29b255d25621/projectFiles/dirctoryFirst/class1.java": "Uploaded",
    "projects/e011bad2-0b57-4ed3-a278-29b255d25621/projectFiles/subdirector/subClass1.java": "Uploaded",
    "projects/e011bad2-0b57-4ed3-a278-29b255d25621/projectFiles/sec/classDirSec.java": "Uploaded"
  },
  "projectId": "e011bad2-0b57-4ed3-a278-29b255d25621"
}
```


<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/SoftDevSix/ArgosFileManager/2)
<!-- Reviewable:end -->


# Test Scenarios for File Manager API

## General Information

These test scenarios validate the core functionality of the File Manager API, focusing on the following endpoints:
1. **List Files**
2. **Get File Content**
3. **Upload Directory**

Each scenario includes a description, steps to reproduce, expected results, and test data (where applicable).

---

## 1. List Files

### Test Case 1.1: Valid Project ID
- **Description**: Verify that the API returns a list of files for a valid `projectId`.
- **Steps**:
  1. Send a `GET` request to `/fileManager/files` with a valid `projectId` parameter.
- **Expected Result**: Status code `200 OK` and a JSON array of file names.
- **Test Data**:
  - `projectId`: `test-project-id`

---

### Test Case 1.2: Invalid Project ID
- **Description**: Verify that the API handles an invalid or nonexistent `projectId`.
- **Steps**:
  1. Send a `GET` request to `/fileManager/files` with an invalid `projectId` parameter.
- **Expected Result**: Status code `404 Not Found` or an appropriate error message.

---

### Test Case 1.3: Missing Project ID Parameter
- **Description**: Verify the response when the `projectId` parameter is missing.
- **Steps**:
  1. Send a `GET` request to `/fileManager/files` without the `projectId` parameter.
- **Expected Result**: Status code `500` with an appropriate error message.

---

## 2. Get File Content

### Test Case 2.1: Valid File Path
- **Description**: Verify that the API returns the content of a file for a valid `projectId` and `filePath`.
- **Steps**:
  1. Send a `GET` request to `/fileManager/file` with valid `projectId` and `filePath` parameters.
- **Expected Result**: Status code `200 OK` and the file content as a string.
- **Test Data**:
  - `projectId`: `test-project-id`
  - `filePath`: `file1.java`

### Test Case 2.3: Missing Parameters
- **Description**: Verify the response when required parameters (`projectId` or `filePath`) are missing.
- **Steps**:
  1. Send a `GET` request to `/fileManager/file` without one or both parameters.
- **Expected Result**: Status code `400 Bad Request` with an appropriate error message.

---

## 3. Upload Directory

### Test Case 3.1: Valid Directory Upload
- **Description**: Verify that files in a valid local directory are uploaded successfully.
- **Steps**:
  1. Send a `POST` request to `/fileManager/upload` with a valid `localDir` parameter.
- **Expected Result**: Status code `200 OK`, with a JSON response showing `Uploaded` status for all files.
- **Test Data**:
  - `localDir`: `test/`

---

### Test Case 3.2: Invalid Directory Path
- **Description**: Verify the response when the `localDir` parameter is invalid or points to a nonexistent directory.
- **Steps**:
  1. Send a `POST` request to `/fileManager/upload` with an invalid `localDir` parameter.
- **Expected Result**: Status code `400 Bad Request` or an appropriate error message.

---

### Test Case 3.3: Empty Directory Upload
- **Description**: Verify the response when the `localDir` is valid but contains no files.
- **Steps**:
  1. Send a `POST` request to `/fileManager/upload` with a valid but empty `localDir`.
- **Expected Result**: Status code `400 Bad Request` or an appropriate error message indicating no files were uploaded.

0santisimo0 committed:
feat: aws Configuration and dependencies added

0santisimo0 committed:
feat: file manager repository, interface added

0santisimo0 committed:
feat: duplicated project upload validation

0santisimo0 committed:
feat: file Manager controller and service

0santisimo0 committed:
feat: directory Upload, getters and env file created

0santisimo0 committed:
feat: javadocs and refactor

0santisimo0 committed:
feat: s3 repository tests and dependencies for test

0santisimo0 committed:
feat: api general tests and javadocs

0santisimo0 committed:
feat: readme update v1

0santisimo0 committed:
feat: multiple projects addition

0santisimo0 committed:
feat: tests update and javadocs

0santisimo0 committed:
feat: coverage tests

0santisimo0 committed:
feat: exceptions handler

0santisimo0 committed:
feat: sonar Tests coverage

0santisimo0 committed:
fix: sonarQube tests to complete coverage pipline

0santisimo0 committed:
feat: formmating and refactor

0santisimo0 committed:
fix: github pipeline update github env variables

0santisimo0 committed:
fix: innecesary comments

fabian-romero-jalafund commented on src/main/java/org/argos/file/manager/service/S3FileService.java:
```diff
@@ -0,0 +1,76 @@
+     * @param storageRepository the repository for interacting with the S3 storage.
+     */
+    public S3FileService(IStorageRepository storageRepository) {
+        this.storageRepository = storageRepository;
```
You can use lombok @Value or @AllArgsConstructor here

0santisimo0 replied:
Done.
<!-- Reviewable comment -OD6Ee6l94a_C_Mi7wov:b-896fix -->
<!-- Sent from Reviewable.io -->

zrock-dev reviewed (changes requested):
Good job. I couldn't run the project yet.

NT: It would be great to use string.format instead of concatenation to improve our performance.

> Reviewed 20 of 22 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-:-OCndwW19QWxCs2DVY5B:b-7jyrbp)* status: 20 of 22 files reviewed, 10 unresolved discussions (waiting on @0santisimo0, @gabygabylagaby, and @zamoranoalicia)

___
*[`src/main/java/org/argos/file/manager/controller/FileManagerController.java` line 23 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCn_UVm4r9HIIGp-ZDk:-OCn_UVm4r9HIIGp-ZDl:bb16t17) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/controller/FileManagerController.java#L23)):*
Major:

Since this has only one constructor, that annotation is not required

> As of Spring Framework 4.3, an `@Autowired` annotation on such a constructor is no longer necessary if the target bean defines only one constructor to begin with. However, if several constructors are available and there is no primary/default constructor, at least one of the constructors must be annotated with `@Autowired` in order to instruct the container which one to use. See the discussion on [constructor resolution](https://docs.spring.io/spring-framework/reference/core/beans/annotation-config/autowired.html#beans-autowired-annotation-constructor-resolution) for details.
https://docs.spring.io/spring-framework/reference/core/beans/annotation-config/autowired.html
___
*[`src/main/java/org/argos/file/manager/ArgosFileManagerApplication.java` line 9 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnZO7w16faPlQMirz_:-OCnZO7x7CQ-G4VHnNzd:bks4fww) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/ArgosFileManagerApplication.java#L9)):*
> ```Java
> public class ArgosFileManagerApplication {
> 
>     @SuppressWarnings("java:S1186")
> ```

Minor

Why is this SonarQube rule being suppressed?
___
*[`src/main/java/org/argos/file/manager/controller/FileManagerController.java` line 13 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCn_0si7SNl51G9TJ9S:-OCn_0si7SNl51G9TJ9T:b3h0n08) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/controller/FileManagerController.java#L13)):*
> ```Java
>  */
> @RestController
> @RequestMapping("/api")
> ```

Minor:

What about using "fileManager" insteaad of "api" ? This will give us context to identify the endpoint.
___
*[`src/test/java/org/argos/file/manager/fileManager/ExceptionClassesTest.java` line 14 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnfCB93vr_hvXcLvXY:-OCnfCB93vr_hvXcLvXZ:b6ateu8) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/test/java/org/argos/file/manager/fileManager/ExceptionClassesTest.java#L14)):*
> ```Java
> import org.springframework.http.ResponseEntity;
> 
> class ExceptionClassesTest {
> ```

What about testing when the exception is thrown? I believe these tests are verifying an exception, but it would be more valuable to test when the exception is thrown. This way we'll have control over this.
___
*[`src/main/java/org/argos/file/manager/repository/S3Repository.java` line 56 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnclDvBTXVuDvaLdQB:-OCnclDvBTXVuDvaLdQC:bd9b71f) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/repository/S3Repository.java#L56)):*
> ```Java
>         if (!Files.exists(directory) || !Files.isDirectory(directory)) {
>             throw new BadRequestError("Invalid local directory: " + localDir);
>         }
> ```

Major:

I understand this conditional is verifying if `localDir` is available or not. But since this method is for uploading the project, when is the bunch of directories being created?



_Code quote:_
```Java
        Path directory = Paths.get(localDir);
        if (!Files.exists(directory) || !Files.isDirectory(directory)) {
            throw new BadRequestError("Invalid local directory: " + localDir);
        }
```
___
*[`src/main/java/org/argos/file/manager/repository/S3Repository.java` line 63 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCndF3m9XwWCZkrVH2w:-OCndF3m9XwWCZkrVH2x:b-karkui) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/repository/S3Repository.java#L63)):*
> ```Java
> 
>             if (files.isEmpty()) {
>                 throw new BadRequestError("No files found in the directory to upload.");
> ```

Major:

What about appending the name of the empty directory? This way the error message would be more informative.
___
*[`src/main/java/org/argos/file/manager/exceptions/GlobalExceptionHandler.java` line 14 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnaiidD_F99SougRqm:-OCnaiie5wZJ3w_8JBE8:b-1bbqwh) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/exceptions/GlobalExceptionHandler.java#L14)):*
> ```Java
>  */
> @RestControllerAdvice
> public class GlobalExceptionHandler {
> ```

Tip:

Would it be possible to also log when an exception is thrown. This way the developer will have access to those errors.
___
*[`src/test/java/org/argos/file/manager/fileManager/FileManagerControllerTest.java` line 34 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnfbPE6GDY2S3SkhKa:-OCnfbPE6GDY2S3SkhKb:b-wb1o97) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/test/java/org/argos/file/manager/fileManager/FileManagerControllerTest.java#L34)):*
> ```Java
>      */
>     @Test
>     void testListFiles() throws Exception {
> ```

Major:

What about catching the exception to verify if is expected or not.
___
*[`src/test/java/org/argos/file/manager/fileManager/FileManagerServiceTest.java` line 40 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCng1kGEm3zFy2CPqTv:-OCng1kGEm3zFy2CPqTw:b-7s6iem) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/test/java/org/argos/file/manager/fileManager/FileManagerServiceTest.java#L40)):*
> ```Java
>      */
>     @Test
>     void testUploadDirectory() {
> ```

Major:

What about uploading an actual java project? This way we can verify that specific behavior of the FileManager service
___
*[`src/main/java/org/argos/file/manager/service/S3FileService.java` line 54 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCneec48b1-aCpTdhZc:-OCneec5-3i01lAWHJbw:b-ibcmy0) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/service/S3FileService.java#L54)):*
> ```Java
>     private String generateProjectId() {
>         return UUID.randomUUID().toString();
>     }
> ```

What about extracting this method to a new class. Because its implementation is more about generating an ID and less about the overall service.



_Code quote:_
```Java
    private String generateProjectId() {
        return UUID.randomUUID().toString();
    }
```


<!-- Sent from Reviewable.io -->

0santisimo0 committed:
fix: github workflow aws variables

0santisimo0 committed:
fix: refactor, validators and comments fixed

0santisimo0 committed:
fix: format violations fixes

0santisimo0 commented:
> Good job. I couldn't run the project yet.
> 
> NT: It would be great to use string.format instead of concatenation to improve our performance.
> 
> > Reviewed 20 of 22 files at r1, 2 of 2 files at r2.
> > _[Reviewable](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-:-OCndwW19QWxCs2DVY5B:b-7jyrbp)_ status: 20 of 22 files reviewed, 10 unresolved discussions (waiting on @0santisimo0, @gabygabylagaby, and @zamoranoalicia)
> 
> _[`src/main/java/org/argos/file/manager/controller/FileManagerController.java` line 23 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCn_UVm4r9HIIGp-ZDk:-OCn_UVm4r9HIIGp-ZDl:bb16t17) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/controller/FileManagerController.java#L23)):_ Major:
> 
> Since this has only one constructor, that annotation is not required

@0santisimo0 : Removed :heavy_check_mark: 

> 
> > As of Spring Framework 4.3, an `@Autowired` annotation on such a constructor is no longer necessary if the target bean defines only one constructor to begin with. However, if several constructors are available and there is no primary/default constructor, at least one of the constructors must be annotated with `@Autowired` in order to instruct the container which one to use. See the discussion on [constructor resolution](https://docs.spring.io/spring-framework/reference/core/beans/annotation-config/autowired.html#beans-autowired-annotation-constructor-resolution) for details.
> > https://docs.spring.io/spring-framework/reference/core/beans/annotation-config/autowired.html
> 
> _[`src/main/java/org/argos/file/manager/ArgosFileManagerApplication.java` line 9 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnZO7w16faPlQMirz_:-OCnZO7x7CQ-G4VHnNzd:bks4fww) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/ArgosFileManagerApplication.java#L9)):_
> 
> > ```java
> > public class ArgosFileManagerApplication {
> > 
> >     @SuppressWarnings("java:S1186")
> > ```
> 
> Minor
> 
> Why is this SonarQube rule being suppressed?

@0santisimo0 : Removed :heavy_check_mark: 

> 
> _[`src/main/java/org/argos/file/manager/controller/FileManagerController.java` line 13 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCn_0si7SNl51G9TJ9S:-OCn_0si7SNl51G9TJ9T:b3h0n08) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/controller/FileManagerController.java#L13)):_
> 
> > ```java
> >  */
> > @RestController
> > @RequestMapping("/api")
> > ```
> 
> Minor:
> 
> What about using "fileManager" insteaad of "api" ? This will give us context to identify the endpoint.

@0santisimo0 : Fixed :heavy_check_mark: 

> 
> _[`src/test/java/org/argos/file/manager/fileManager/ExceptionClassesTest.java` line 14 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnfCB93vr_hvXcLvXY:-OCnfCB93vr_hvXcLvXZ:b6ateu8) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/test/java/org/argos/file/manager/fileManager/ExceptionClassesTest.java#L14)):_
> 
> > ```java
> > import org.springframework.http.ResponseEntity;
> > 
> > class ExceptionClassesTest {
> > ```
> 
> What about testing when the exception is thrown? I believe these tests are verifying an exception, but it would be more valuable to test when the exception is thrown. This way we'll have control over this.

---

@0santisimo0 : Invalid Comment

In the `S3RepositoryTest` file, we have already tested the scenarios where exceptions are thrown. Specifically, the following test cases ensure that exceptions are thrown and handled correctly:

- **`testGetFileContent_ThrowsIOException`**: Verifies that an `IOException` is thrown when there is an issue retrieving the file content.
- **`testGetFileContent_EmptyFilePath_ThrowsBadRequestError`**: Tests that a `BadRequestError` is thrown when the file path is empty.
- **`testGetFileContent_EmptyProjectId_ThrowsBadRequestError`**: Ensures that a `BadRequestError` is thrown when the project ID is empty.
- **`testGetFileContent_ThrowsNotFoundError`**: Verifies that a `NotFoundError` is thrown when the file cannot be found.
- **`testListFiles_ThrowsBadRequestError`**: Checks that a `BadRequestError` is thrown when there is an issue listing files in S3.
- **`testListFiles_EmptyBucket`**: Ensures the appropriate response is returned when the bucket is empty (though no exception is thrown in this case).

---

> 
> _[`src/main/java/org/argos/file/manager/repository/S3Repository.java` line 56 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnclDvBTXVuDvaLdQB:-OCnclDvBTXVuDvaLdQC:bd9b71f) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/repository/S3Repository.java#L56)):_
> 
> > ```java
> >         if (!Files.exists(directory) || !Files.isDirectory(directory)) {
> >             throw new BadRequestError("Invalid local directory: " + localDir);
> >         }
> > ```
> 
> Major:
> 
> I understand this conditional is verifying if `localDir` is available or not. But since this method is for uploading the project, when is the bunch of directories being created?

@0santisimo0 :  This method just verifies if the directory uploaded exists and if it is a directory   :heavy_check_mark: 

> 
> _Code quote:_
> 
> ```java
>         Path directory = Paths.get(localDir);
>         if (!Files.exists(directory) || !Files.isDirectory(directory)) {
>             throw new BadRequestError("Invalid local directory: " + localDir);
>         }
> ```
> 
> _[`src/main/java/org/argos/file/manager/repository/S3Repository.java` line 63 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCndF3m9XwWCZkrVH2w:-OCndF3m9XwWCZkrVH2x:b-karkui) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/repository/S3Repository.java#L63)):_
> 
> > ```java
> >             if (files.isEmpty()) {
> >                 throw new BadRequestError("No files found in the directory to upload.");
> > ```
> 
> Major:
> 
> What about appending the name of the empty directory? This way the error message would be more informative.

@0santisimo0 : valid enhacement :heavy_check_mark:

> 
> _[`src/main/java/org/argos/file/manager/exceptions/GlobalExceptionHandler.java` line 14 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnaiidD_F99SougRqm:-OCnaiie5wZJ3w_8JBE8:b-1bbqwh) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/exceptions/GlobalExceptionHandler.java#L14)):_
> 
> > ```java
> >  */
> > @RestControllerAdvice
> > public class GlobalExceptionHandler {
> > ```
> 
> Tip:
> 
> Would it be possible to also log when an exception is thrown. This way the developer will have access to those errors.

@0santisimo0 : semi-valid enhacement :heavy_check_mark:
It is **not necessary**. Logging every exception can lead to several issues. First, it can cause **redundant logging**, as exceptions may already be logged in the service or controller layers, which can lead to **log clutter**. Additionally, logging **every exception**—especially in high-traffic applications—can negatively affect **performance**. It also raises security concerns, as **sensitive information** might be exposed in logs if not handled carefully.


> 
> _[`src/test/java/org/argos/file/manager/fileManager/FileManagerControllerTest.java` line 34 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnfbPE6GDY2S3SkhKa:-OCnfbPE6GDY2S3SkhKb:b-wb1o97) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/test/java/org/argos/file/manager/fileManager/FileManagerControllerTest.java#L34)):_
> 
> > ```java
> >      */
> >     @Test
> >     void testListFiles() throws Exception {
> > ```
> 
> Major:
> 
> What about catching the exception to verify if is expected or not.
> 
> _[`src/test/java/org/argos/file/manager/fileManager/FileManagerServiceTest.java` line 40 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCng1kGEm3zFy2CPqTv:-OCng1kGEm3zFy2CPqTw:b-7s6iem) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/test/java/org/argos/file/manager/fileManager/FileManagerServiceTest.java#L40)):_
> 
> > ```java
> >      */
> >     @Test
> >     void testUploadDirectory() {
> > ```
> 
> Major:
> 
> What about uploading an actual java project? This way we can verify that specific behavior of the FileManager service

@0santisimo0: **Testers' Concerns** - The project upload functionality is already mocked in the tests.

> 
> _[`src/main/java/org/argos/file/manager/service/S3FileService.java` line 54 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCneec48b1-aCpTdhZc:-OCneec5-3i01lAWHJbw:b-ibcmy0) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/service/S3FileService.java#L54)):_
> 
> > ```java
> >     private String generateProjectId() {
> >         return UUID.randomUUID().toString();
> >     }
> > ```
> 
> What about extracting this method to a new class. Because its implementation is more about generating an ID and less about the overall service.

@0santisimo0 : Fixed :heavy_check_mark:

> 
> _Code quote:_
> 
> ```java
>     private String generateProjectId() {
>         return UUID.randomUUID().toString();
>     }
> ```

fabian-romero-jalafund commented on src/main/java/org/argos/file/manager/controller/FileManagerController.java:
```diff
@@ -0,0 +1,61 @@
+     * @param projectId the ID of the project.
+     * @return a list of file keys in the S3 bucket for the given project.
+     */
+    @GetMapping("/files")
```
add CORS to your endpoints please to allow make petitions from the front end
 ```@CrossOrigin(origins = "http://localhost:5173")```

0santisimo0 replied:
![image](https://github.com/user-attachments/assets/61f74200-8682-41da-9e84-c1af07ef261b)
Solved :heavy_check_mark:

GabyGV-GGV commented on .github/workflows/check-push.yml:
```diff
@@ -23,6 +28,12 @@ jobs:
+      - name: Debug AWS Environment Variables
+        run: |
+          echo "AWS_REGION=$AWS_REGION"
+          echo "AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID:0:4}******"  # Redact key for security
```
there is no need for such comments...

0santisimo0 replied:
INVALID: The `echo` commands are included to debug the AWS environment variables during workflow execution. They can be helpful for troubleshooting if any issues arise in the future.
<!-- Reviewable comment -OCsFafa7z5yFjG_K1M3:bt6psp4 -->
<!-- Sent from Reviewable.io -->

GabyGV-GGV commented on .github/workflows/sonar-scan.yml:
```diff
@@ -9,18 +9,31 @@ jobs:
     name: Build and Sonarqube Analyze
     runs-on: ubuntu-latest
+
+    env:  # Add AWS environment variables here
```
here the same

0santisimo0 replied:
INVALID: The `echo` commands are included to debug the AWS environment variables during workflow execution. They can be helpful for troubleshooting if any issues arise in the future.
<!-- Reviewable comment -OD6DYEoDPUviUG_6TVf:bt6psp4 -->
<!-- Sent from Reviewable.io -->

GabyGV-GGV commented on .github/workflows/sonar-scan.yml:
```diff
@@ -9,18 +9,31 @@ jobs:
+      - name: Debug AWS Environment Variables
+        run: |
+          echo "AWS_REGION=$AWS_REGION"
+          echo "AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID:0:4}******"  # Redact key for security
```
here the same

0santisimo0 replied:
INVALID: The `echo` commands are included to debug the AWS environment variables during workflow execution. They can be helpful for troubleshooting if any issues arise in the future.
<!-- Reviewable comment -OD6D_1-9Av8-86r1tYb:bt6psp4 -->
<!-- Sent from Reviewable.io -->

GabyGV-GGV commented on gradle/libs.versions.toml:
```diff
@@ -17,7 +18,8 @@ lombok = { id = "io.freefair.lombok", version = "8.11" }
 junit-platform-launcher = { module = "org.junit.platform:junit-platform-launcher" }
 lombok = { module = "org.projectlombok:lombok", version.ref = "lombok" }
+dotenv-java = { module = "io.github.cdimascio:dotenv-java", version = "3.0.0" }
+aws-s3 = { module = "software.amazon.awssdk:s3", version = "2.20.108" }
```
do not delete the final line or it will look like you deleted something.

0santisimo0 replied:
so?

0santisimo0 committed:
fix: innecesary comments deleted

zrock-dev reviewed (dismissed):
Good job. This PR is important and many of my commentaries are nitpicking. I suggest to create a follow up PR addressing those. 

There is only one major or important change.

> Reviewed 15 of 18 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-:-OCpDQK7D7aw8oktXkZO:b-iyhrgh)* status: all files reviewed, 20 unresolved discussions (waiting on @0santisimo0, @fabian-romero-jalafund, @gabygabylagaby, and @zamoranoalicia)

___
*[`src/test/java/org/argos/file/manager/ArgosFileManagerApplicationTests.java` line 0 at r5](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCpCfoqEOi1oF7qeZks:-OCpCfoqEOi1oF7qeZkt:b-rf3moj) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/4d710af50af4d682d3b52f37d38bb64bf64e7001/src/test/java/org/argos/file/manager/ArgosFileManagerApplicationTests.java#L0)):*
Major:

This test was verifying if it can load the spring context. Has this test been deleted because the loadContext is being tested somewhere else?
___
*[`src/main/java/org/argos/file/manager/utils/InputValidator.java` line 10 at r5](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCpCIzkEpH4XUv2zGTa:-OCpCIzl4Ah6bJzdidDk:bmv9hb6) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/4d710af50af4d682d3b52f37d38bb64bf64e7001/src/main/java/org/argos/file/manager/utils/InputValidator.java#L10)):*
> ```Java
>  * This class validates input such as project ID, directory paths, and file paths.
>  */
> @SuppressWarnings("java:S6548")
> ```

Nit:

What rule is being suppressed?
___
*[`build.gradle.kts` line 106 at r5](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCpB5AR4VjmqqyZUnxT:-OCpB5AR4VjmqqyZUnxU:b-5x4glu) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/4d710af50af4d682d3b52f37d38bb64bf64e7001/build.gradle.kts#L106)):*
> ```Kotlin
>         property("sonar.qualitygate.wait", "true")
>         property("sonar.exclusions", "**/S3Config.java")
>         property("sonar.exclusions", "**/ArgosFileManagerApplication.java")
> ```

Nit:

Why should SonarQube ignore these files?



_Code quote:_
```Kotlin
        property("sonar.exclusions", "**/S3Config.java")
        property("sonar.exclusions", "**/ArgosFileManagerApplication.java")
```
___
*[`.github/workflows/check-pull-request.yml` line 36 at r5](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCpBOw0-gFAOo-hVrPM:-OCpBOw0-gFAOo-hVrPN:b9p5gaa) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/4d710af50af4d682d3b52f37d38bb64bf64e7001/.github/workflows/check-pull-request.yml#L36)):*
> ```YAML
>           echo "AWS_REGION=$AWS_REGION"
>           echo "AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID:0:4}******"
>         shell: bash
> ```

Nit:
Once it's running, these lines won't be required.



_Code quote:_
```YAML
      - name: Debug AWS Environment Variables
        run: |
          echo "AWS_REGION=$AWS_REGION"
          echo "AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID:0:4}******"
        shell: bash
```
___
*[`src/main/java/org/argos/file/manager/utils/S3KeyGenerator.java` line 9 at r5](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCpCSFBCKwK2_1TFfzR:-OCpCSFBCKwK2_1TFfzS:b-tdlg7f) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/4d710af50af4d682d3b52f37d38bb64bf64e7001/src/main/java/org/argos/file/manager/utils/S3KeyGenerator.java#L9)):*
> ```Java
>  * This class cannot be instantiated as it contains only static methods.
>  */
> public class S3KeyGenerator {
> ```

Nit:

It would be just a KeyGenerator.Because it doesn't have anything to do with S3


<!-- Sent from Reviewable.io -->

0santisimo0 committed:
fix: rename class

0santisimo0 reviewed (commented):
> Reviewed 9 of 22 files at r1, 15 of 18 files at r4, 3 of 3 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/softdevsix/argosfilemanager/2)* status: all files reviewed, 17 unresolved discussions (waiting on @AllArgsConstructor, @fabian-romero-jalafund, @gabygabylagaby, @value, @zamoranoalicia, and @zrock-dev)

___
*[`build.gradle.kts` line 106 at r5](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCpB5AR4VjmqqyZUnxT:-OD68piI0eFYGMgMW2NE:bpjfgo9) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/4d710af50af4d682d3b52f37d38bb64bf64e7001/build.gradle.kts#L106)):*
<details><summary><i>Previously, zrock-dev (Santiago C. Saavedra) wrote…</i></summary><blockquote>

Nit:

Why should SonarQube ignore these files?
</blockquote></details>

S3config has to be tested from enviorment file and Tests have issues while acceding to Github Secrets to test the connection
ArgosFileManagerApplication.- I removed it from test because it is the one that when testing initialized the application and requested bean in all classes, it gave locally but in the github build there were problems
___
*[`.github/workflows/check-pull-request.yml` line 36 at r5](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCpBOw0-gFAOo-hVrPM:-OD6CPLc7YvtXURRoxdP:bt6psp4) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/4d710af50af4d682d3b52f37d38bb64bf64e7001/.github/workflows/check-pull-request.yml#L36)):*
<details><summary><i>Previously, zrock-dev (Santiago C. Saavedra) wrote…</i></summary><blockquote>

Nit:
Once it's running, these lines won't be required.
</blockquote></details>

INVALID: The `echo` commands are included to debug the AWS environment variables during workflow execution. They can be helpful for troubleshooting if any issues arise in the future.
___
*[`src/main/java/org/argos/file/manager/service/S3FileService.java` line 54 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCneec48b1-aCpTdhZc:-OD6LQHa9S5R2gDRgVgr:b-886gsx) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/main/java/org/argos/file/manager/service/S3FileService.java#L54)):*
<details><summary><i>Previously, 0santisimo0 (Santiago Caballero) wrote…</i></summary><blockquote>

> > ```java
> >     private String generateProjectId() {
> >         return UUID.randomUUID().toString();
> >     }
> > ```
> > 

> 
> What about extracting this method to a new class. Because its implementation is more about generating an ID and less about the overall service.
> 

@0santisimo0 : Fixed :heavy_check_mark:


> 
> _Code quote:_
> 
> ```java
>     private String generateProjectId() {
>         return UUID.randomUUID().toString();
>     }
> ```
> 
</blockquote></details>

Fixed
___
*[`src/main/java/org/argos/file/manager/utils/InputValidator.java` line 10 at r5](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCpCIzkEpH4XUv2zGTa:-OD6EjvK3VMBplh4y5NS:b-iraltp) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/4d710af50af4d682d3b52f37d38bb64bf64e7001/src/main/java/org/argos/file/manager/utils/InputValidator.java#L10)):*
<details><summary><i>Previously, zrock-dev (Santiago C. Saavedra) wrote…</i></summary><blockquote>

Nit:

What rule is being suppressed?
</blockquote></details>

SonarQube, by default, flags the use of the Singleton design pattern as a potential issue as default. However, the decision to use the Singleton pattern depends on the specific requirements of the project. In this case, I have chosen to use the Singleton pattern intentionally and suppressed the alert
___
*[`src/test/java/org/argos/file/manager/ArgosFileManagerApplicationTests.java` line  at r5](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCpCfoqEOi1oF7qeZks:-OD6GOEzCzBN2w8Ga_cl:b-b0h854) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/4d710af50af4d682d3b52f37d38bb64bf64e7001/src/test/java/org/argos/file/manager/ArgosFileManagerApplicationTests.java#L)):*
<details><summary><i>Previously, zrock-dev (Santiago C. Saavedra) wrote…</i></summary><blockquote>

Major:

This test was verifying if it can load the spring context. Has this test been deleted because the loadContext is being tested somewhere else?
</blockquote></details>

The `loadContext` test was not necessary in this pull request because context loading is not the focus here. Additionally, testing the context in this way triggers the initialization of the entire application, expecting beans in all classes. This was causing the build to break, so the test was removed to prevent unnecessary initialization and ensure the build remains stable.
___
*[`src/test/java/org/argos/file/manager/fileManager/ExceptionClassesTest.java` line 14 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnfCB93vr_hvXcLvXY:-OD6I27yC8hGZcKRRjRn:btx03ri) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/test/java/org/argos/file/manager/fileManager/ExceptionClassesTest.java#L14)):*
<details><summary><i>Previously, 0santisimo0 (Santiago Caballero) wrote…</i></summary><blockquote>

---

@0santisimo0 : Invalid Comment

In the `S3RepositoryTest` file, we have already tested the scenarios where exceptions are thrown. Specifically, the following test cases ensure that exceptions are thrown and handled correctly:

- **`testGetFileContent_ThrowsIOException`**: Verifies that an `IOException` is thrown when there is an issue retrieving the file content.
- **`testGetFileContent_EmptyFilePath_ThrowsBadRequestError`**: Tests that a `BadRequestError` is thrown when the file path is empty.
- **`testGetFileContent_EmptyProjectId_ThrowsBadRequestError`**: Ensures that a `BadRequestError` is thrown when the project ID is empty.
- **`testGetFileContent_ThrowsNotFoundError`**: Verifies that a `NotFoundError` is thrown when the file cannot be found.
- **`testListFiles_ThrowsBadRequestError`**: Checks that a `BadRequestError` is thrown when there is an issue listing files in S3.
- **`testListFiles_EmptyBucket`**: Ensures the appropriate response is returned when the bucket is empty (though no exception is thrown in this case).

---

</blockquote></details>

 Invalid Comment

In the `S3RepositoryTest` file, we have already tested the scenarios where exceptions are thrown. Specifically, the following test cases ensure that exceptions are thrown and handled correctly:

-   **`testGetFileContent_ThrowsIOException`**: Verifies that an `IOException` is thrown when there is an issue retrieving the file content.
-   **`testGetFileContent_EmptyFilePath_ThrowsBadRequestError`**: Tests that a `BadRequestError` is thrown when the file path is empty.
-   **`testGetFileContent_EmptyProjectId_ThrowsBadRequestError`**: Ensures that a `BadRequestError` is thrown when the project ID is empty.
-   **`testGetFileContent_ThrowsNotFoundError`**: Verifies that a `NotFoundError` is thrown when the file cannot be found.
-   **`testListFiles_ThrowsBadRequestError`**: Checks that a `BadRequestError` is thrown when there is an issue listing files in S3.
-   **`testListFiles_EmptyBucket`**: Ensures the appropriate response is returned when the bucket is empty (though no exception is thrown in this case).
___
*[`src/test/java/org/argos/file/manager/fileManager/FileManagerControllerTest.java` line 34 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCnfbPE6GDY2S3SkhKa:-OD6KmFV2penvMlAct06:bxvpala) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/test/java/org/argos/file/manager/fileManager/FileManagerControllerTest.java#L34)):*
<details><summary><i>Previously, zrock-dev (Santiago C. Saavedra) wrote…</i></summary><blockquote>

Major:

What about catching the exception to verify if is expected or not.
</blockquote></details>

There are already tests for exceptions:
- testUploadDirectory\_ThrowsInternalServerError
- testListFiles\_ThrowsBadRequestError
- testGetFileContent\_ThrowsNotFoundError
- testGetFileContent\_EmptyProjectId\_ThrowsBadRequestError
- testGetFileContent\_EmptyFilePath\_ThrowsBadRequestError

etc...
___
*[`src/test/java/org/argos/file/manager/fileManager/FileManagerServiceTest.java` line 40 at r2](https://reviewable.io/reviews/softdevsix/argosfilemanager/2#-OCng1kGEm3zFy2CPqTv:-OD6LH7zB_RcGFqWxHwf:b341iw) ([raw file](https://github.com/softdevsix/argosfilemanager/blob/ade07513934cbc10c983013d5fa6f14c714a59e1/src/test/java/org/argos/file/manager/fileManager/FileManagerServiceTest.java#L40)):*
<details><summary><i>Previously, 0santisimo0 (Santiago Caballero) wrote…</i></summary><blockquote>

@0santisimo0: **Testers' Concerns** - The project upload functionality is already mocked in the tests.
</blockquote></details>

**Testers' Concerns** - The project upload functionality is already mocked in the tests.


<!-- Sent from Reviewable.io -->
